File: packages/sdk/src/core/managers/device-manager.ts
import type { HttpClient } from '../client/http-client';
import type { OfflineQueue } from '../queue/offline-queue';
import { getItem, setItem } from '../storage/storage';
import type { CreateDeviceRequest, DeviceInfo, PhaseConfig } from '../types';
import { STORAGE_KEYS } from '../types';
import { generateDeviceId } from '../utils/id-generator';
import { logger } from '../utils/logger';
import { validateDeviceId } from '../utils/validator';

export class DeviceManager {
  private deviceId: string | null = null;
  private initPromise: Promise<string> | null = null;
  private readonly httpClient: HttpClient;
  private readonly offlineQueue: OfflineQueue;
  private readonly getDeviceInfo: () => DeviceInfo;
  private readonly collectDeviceInfo: boolean;
  private readonly collectLocale: boolean;

  constructor(
    httpClient: HttpClient,
    offlineQueue: OfflineQueue,
    getDeviceInfo: () => DeviceInfo,
    config?: PhaseConfig
  ) {
    this.httpClient = httpClient;
    this.offlineQueue = offlineQueue;
    this.getDeviceInfo = getDeviceInfo;
    this.collectDeviceInfo = config?.deviceInfo ?? true;
    this.collectLocale = config?.userLocale ?? true;
  }

  async initialize(): Promise<string> {
    if (this.initPromise) {
      return this.initPromise;
    }
    if (this.deviceId) {
      return this.deviceId;
    }

    this.initPromise = this.doInitialize();

    try {
      return await this.initPromise;
    } finally {
      this.initPromise = null;
    }
  }

  private async doInitialize(): Promise<string> {
    const result = await getItem<string>(STORAGE_KEYS.DEVICE_ID);
    const stored = result.success ? result.data : null;

    if (stored) {
      const validation = validateDeviceId(stored);
      if (validation.success) {
        this.deviceId = stored;
      } else {
        logger.error('Stored device ID invalid, generating new one');
        this.deviceId = generateDeviceId();
        try {
          await setItem(STORAGE_KEYS.DEVICE_ID, this.deviceId);
        } catch {
          logger.error('Failed to persist device ID');
        }
      }
    } else {
      this.deviceId = generateDeviceId();
      try {
        await setItem(STORAGE_KEYS.DEVICE_ID, this.deviceId);
      } catch {
        logger.error('Failed to persist device ID');
      }
    }

    return this.deviceId;
  }

  async identify(isOnline: boolean): Promise<void> {
    if (!this.deviceId) {
      logger.error('Device ID not set, call initialize() first');
      return;
    }

    await this.registerDevice(isOnline);
  }

  getDeviceId(): string | null {
    return this.deviceId;
  }

  private buildDevicePayload(): CreateDeviceRequest | null {
    if (!this.deviceId) {
      return null;
    }

    const deviceInfo = this.getDeviceInfo();
    return {
      deviceId: this.deviceId,
      deviceType: this.collectDeviceInfo ? deviceInfo.deviceType : null,
      osVersion: this.collectDeviceInfo ? deviceInfo.osVersion : null,
      platform: this.collectDeviceInfo ? deviceInfo.platform : null,
      locale: this.collectLocale ? deviceInfo.locale : null,
      disableGeolocation: !this.collectLocale,
    };
  }

  private async registerDevice(isOnline: boolean): Promise<void> {
    const payload = this.buildDevicePayload();
    if (!payload) {
      logger.error('Device ID not set, cannot register device');
      return;
    }

    if (isOnline) {
      try {
        const result = await this.httpClient.createDevice(payload);
        if (result.success) {
          await this.cacheDeviceInfo(payload);
        } else {
          logger.error('Failed to register device, queueing', result.error);
          try {
            await this.offlineQueue.enqueue({ type: 'device', payload });
          } catch (error) {
            logger.error('Failed to queue device registration', error);
          }
        }
      } catch (error) {
        logger.error('Exception during device registration, queueing', error);
        try {
          await this.offlineQueue.enqueue({ type: 'device', payload });
        } catch (queueError) {
          logger.error('Failed to queue device registration', queueError);
        }
      }
    } else {
      try {
        await this.offlineQueue.enqueue({ type: 'device', payload });
      } catch (error) {
        logger.error('Failed to queue device registration', error);
      }
    }
  }

  private async cacheDeviceInfo(payload: CreateDeviceRequest): Promise<void> {
    try {
      await setItem(STORAGE_KEYS.DEVICE_INFO, payload);
      logger.debug('Device info cached successfully');
    } catch {
      logger.error('Failed to cache device info');
    }
  }
}

File: packages/swift/Sources/PhaseAnalytics/Core/Managers/DeviceManager.swift
import Foundation

internal actor DeviceManager {
    private var deviceID: String?
    private let httpClient: HTTPClient
    private let offlineQueue: OfflineQueue
    private let storage: StorageAdapter
    private let getDeviceInfo: @Sendable () -> DeviceInfo
    private let collectDeviceInfo: Bool
    private let collectLocale: Bool

    private var initPromise: Task<String, Never>?

    init(
        httpClient: HTTPClient,
        offlineQueue: OfflineQueue,
        storage: StorageAdapter,
        getDeviceInfo: @escaping @Sendable () -> DeviceInfo,
        config: PhaseConfig
    ) {
        self.httpClient = httpClient
        self.offlineQueue = offlineQueue
        self.storage = storage
        self.getDeviceInfo = getDeviceInfo
        self.collectDeviceInfo = config.deviceInfo
        self.collectLocale = config.userLocale
    }

    func initialize(isOnline: Bool) async -> String {
        if let promise = initPromise {
            return await promise.value
        }

        if let existingID = deviceID {
            return existingID
        }

        let promise = Task<String, Never> {
            await doInitialize(isOnline: isOnline)
        }
        initPromise = promise

        let result = await promise.value
        initPromise = nil
        return result
    }

    private func doInitialize(isOnline: Bool) async -> String {
        let result = await storage.getItem(key: StorageKeys.deviceID) as Result<String?, Error>

        if case .success(let stored) = result,
            let stored = stored,
            case .success = Validator.validateDeviceID(stored)
        {
            deviceID = stored
        } else {
            deviceID = IDGenerator.generateDeviceID()

            let saveResult = await storage.setItem(key: StorageKeys.deviceID, value: deviceID!)
            if case .failure(let error) = saveResult {
                logger.error("Failed to persist device ID. Storage unavailable.", error)
            }
        }

        return deviceID!
    }

    func identify(isOnline: Bool, properties: DeviceProperties? = nil) async {
        guard deviceID != nil else {
            logger.error("Device ID not set. Call initialize() first.")
            return
        }

        await registerDevice(isOnline: isOnline, properties: properties, force: true)
    }

    private func shouldUpdateDevice() async -> Bool {
        guard let current = buildDevicePayload() else {
            logger.error("Device ID not set. Cannot check for updates.")
            return false
        }

        let result = await storage.getItem(key: StorageKeys.deviceInfo) as Result<CreateDeviceRequest?, Error>

        if case .success(let cached) = result, let cached = cached {
            let hasChanged =
                cached.osVersion != current.osVersion
                || cached.platform != current.platform || cached.locale != current.locale
                || cached.model != current.model
                || cached.disableGeolocation != current.disableGeolocation

            return hasChanged
        }

        return true
    }

    private func registerDevice(isOnline: Bool, properties: DeviceProperties? = nil, force: Bool = false) async {
        guard let payload = buildDevicePayload(properties: properties) else {
            logger.error("Device ID not set. Cannot register device.")
            return
        }

        if isOnline {
            let result = await httpClient.createDevice(payload)

            if case .success = result {
                await cacheDeviceInfo(payload)
            } else if case .failure(let error) = result {
                logger.error("Device registration failed. Queuing for retry.", error)
                await offlineQueue.enqueue(.device(payload: payload, clientOrder: 0, retryCount: nil))
                await cacheDeviceInfo(payload)
            }
        } else {
            await offlineQueue.enqueue(.device(payload: payload, clientOrder: 0, retryCount: nil))
            await cacheDeviceInfo(payload)
        }
    }

    private func cacheDeviceInfo(_ payload: CreateDeviceRequest) async {
        let result = await storage.setItem(key: StorageKeys.deviceInfo, value: payload)
        if case .success = result {
            logger.info("Device info cached successfully")
        } else if case .failure(let error) = result {
            logger.error("Failed to cache device info", error)
        }
    }

    private func buildDevicePayload(properties: DeviceProperties? = nil) -> CreateDeviceRequest? {
        guard let deviceID = deviceID else { return nil }

        let info = getDeviceInfo()
        let propsDict: [String: AnyCodable]? = properties?.dictionary.mapValues { AnyCodable($0) }

        return CreateDeviceRequest(
            deviceId: deviceID,
            osVersion: collectDeviceInfo ? info.osVersion : nil,
            platform: collectDeviceInfo ? info.platform : nil,
            locale: collectLocale ? info.locale : nil,
            model: collectDeviceInfo ? info.model : nil,
            properties: propsDict,
            disableGeolocation: !collectLocale
        )
    }
}
